### PR TITLE
Java: Model Guava collections package 

### PR DESCRIPTION
**Pull request by joefarebrother**
~~Depends on https://github.com/github/codeql/pull/5796~~

### CODE REVIEW

File: java/ql/src/semmle/code/java/frameworks/guava/Collections.qll [hunk 1]
```diff
@@ -3,104 +3,534 @@
...
+    row =
+      [
+        //"package;type;overrides;name;signature;ext;inputspec;outputspec;kind",
+        // Multisets
+        "com.google.common.collect;Multiset;true;add;(Object,int);;Argument[0];Element of Argument[-1];value",
+        "com.google.common.collect;Multiset;true;setCount;(Object,int);;Argument[0];Element of Argument[-1];value",
+        "com.google.common.collect;Multiset;true;setCount;(Object,int,int);;Argument[0];Element of Argument[-1];value",
+        "com.google.common.collect;Multiset;true;elementSet;();;Element of Argument[-1];Element of ReturnValue;value",
+        "com.google.common.collect;Multiset;true;entrySet;();;Element of Argument[-1];Element of Element of ReturnValue;value",
+        "com.google.common.collect;Multiset<>$Entry;true;getElement;();;Element of Argument[-1];ReturnValue;value",
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/guava/Collections.qll:22]: All of these `<>` qualifiers should no longer be required

Comment (joefarebrother) [reply java/ql/src/semmle/code/java/frameworks/guava/Collections.qll:22]: Could you rebase https://github.com/github/codeql/pull/5796 onto main so I can more easily make this change?

Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/guava/Collections.qll:22]: Done

Comment (joefarebrother) [reply java/ql/src/semmle/code/java/frameworks/guava/Collections.qll:22]: Hmm; I don't seem to be able to get this change working; removing the `<>`s makes the tests fail. Additionally one of the tests that depends on `Map.Entry` is also failing. It seems like the change that should have made the `<>`s unneeded hasn't actually worked.

Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/guava/Collections.qll:22]: It depends on the new extractor: make sure you're running codeql 2.5.6

Comment (joefarebrother) [reply java/ql/src/semmle/code/java/frameworks/guava/Collections.qll:22]: Ah, thanks

File: java/ql/src/semmle/code/java/frameworks/guava/Collections.qll [hunk 2]
```diff
@@ -3,102 +3,567 @@
...
+        "com.google.common.collect;Multimaps;false;filterValues;(SetMultimap,Predicate);;MapKey of Argument[0];MapKey of ReturnValue;value",
+        "com.google.common.collect;Multimaps;false;filterValues;(SetMultimap,Predicate);;MapValue of Argument[0];MapValue of ReturnValue;value",
+        "com.google.common.collect;Multimaps;false;forMap;(Map);;MapKey of Argument[0];MapKey of ReturnValue;value",
+        "com.google.common.collect;Multimaps;false;forMap;(Map);;MapValue of Argument[0];MapValue of ReturnValue;value",
+        "com.google.common.collect;Multimaps;false;index;(Iterable,Function);;Element of Argument[0];MapValue of ReturnValue;value",
+        "com.google.common.collect;Multimaps;false;index;(Iterator,Function);;Element of Argument[0];MapValue of ReturnValue;value",
+        "com.google.common.collect;Multimaps;false;invertFrom;(Multimap,Multimap);;MapKey of Argument[0];MapValue of ReturnValue;value",
+        "com.google.common.collect;Multimaps;false;invertFrom;(Multimap,Multimap);;MapValue of Argument[0];MapKey of ReturnValue;value",
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/guava/Collections.qll:506]: Target arg1 instead, and add a value edge arg1 -> ret

File: java/ql/src/semmle/code/java/frameworks/guava/Collections.qll [hunk 3]
```diff
@@ -3,102 +3,567 @@
...
+        "com.google.common.collect;Tables;false;transpose;(Table);;SyntheticField[com.google.common.collect.Table.rowKey] of Argument[0];SyntheticField[com.google.common.collect.Table.columnKey] of ReturnValue;value",
+        "com.google.common.collect;Tables;false;transpose;(Table);;SyntheticField[com.google.common.collect.Table.columnKey] of Argument[0];SyntheticField[com.google.common.collect.Table.rowKey] of ReturnValue;value",
+        "com.google.common.collect;Tables;false;transpose;(Table);;MapValue of Argument[0];MapValue of ReturnValue;value",
+        "com.google.common.collect;Tables;false;transformValues;(Table,Function);;SyntheticField[com.google.common.collect.Table.rowKey] of Argument[0];SyntheticField[com.google.common.collect.Table.rowKey] of ReturnValue;value",
+        "com.google.common.collect;Tables;false;transformValues;(Table,Function);;SyntheticField[com.google.common.collect.Table.columnKey] of Argument[0];SyntheticField[com.google.common.collect.Table.columnKey] of ReturnValue;value",
+        "com.google.common.collect;Tables;false;transformValues;(Table,Function);;MapValue of Argument[0];MapValue of ReturnValue;value",
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/guava/Collections.qll:550]: transformValues doesn't necessarily convey taint across values, depends on the lambda passed

File: java/ql/test/library-tests/frameworks/guava/generated/collect/test.ql [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import semmle.code.java.dataflow.ExternalFlow
+import semmle.code.java.dataflow.TaintTracking
+import TestUtilities.InlineExpectationsTest
+
+class SummaryModelTest extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        //"package;type;overrides;name;signature;ext;inputspec;outputspec;kind",
+        "generatedtest;Test;false;newWithElement;;;Argument[0];Element of ReturnValue;value",
```
Comment (smowton) [inline java/ql/test/library-tests/frameworks/guava/generated/collect/test.ql:12]: Don't we have enough models to replace these magic functions with real ones?

**Comment by smowton** [comment]
Performance evaluation at https://github.com/dsp-testing/smowton-dca/issues/3 looks good (Geode apparent result changes investigated and confirmed spurious, and pig, hadoop-common, commons-io and elasticsearch are all Guava users which show no significant slowdown)

**Review by smowton — DISMISSED** [review]

**Review by smowton — APPROVED** [review]

**Pull request merged**